src/proxy-worker.ts

```diff
@@ -0,0 +1,286 @@
+			const { uuid, port, auth } = this.parseRequest(req);
+			this.logger.info(`tunnel requested to device ${uuid} on port ${port}`);
+
+			// we need to use VPN_SERVICE_API_KEY here as this could be an unauthenticated request
+			return device
+				.getDeviceByUUID(uuid, VPN_SERVICE_API_KEY)
+				.tap(data => {
+					if (data == null) {
+						cltSocket.end('HTTP/1.0 404 Not Found\r\n\r\n');
+						throw new errors.HandledTunnelingError(`device not found: ${uuid}`);
+					}
+				})
+				.tap(data =>
+					device
+						.canAccessDevice(data, port, auth)
+						.tap(isAllowed => {
+							if (!isAllowed) {
+								cltSocket.end(
+									'HTTP/1.0 407 Proxy Authorization Required\r\n\r\n',
+								);
+								throw new errors.HandledTunnelingError(
+									`device not accessible: ${uuid}`,
+								);
+							}
+						})
+						.tap(() => {
+							if (!data.is_connected_to_vpn) {
+								cltSocket.end('HTTP/1.0 503 Service Unavailable\r\n\r\n');
+								throw new errors.HandledTunnelingError(
+									`device not available: ${uuid}`,
+								);
+							}
+						}),
+				)
+				.tap(() => (req.url = `${uuid}.vpn:${port}`));
+		})
+			.then(() => next())
+			.catch(errors.APIError, err => {
+				this.logger.alert(`Invalid Response from API (${err.message})`);
+				cltSocket.end('HTTP/1.0 500 Internal Server Error\r\n\r\n');
+			})
+			.catch(errors.BadRequestError, () =>
+				cltSocket.end('HTTP/1.0 400 Bad Request\r\n\r\n'),
+			)
+			.catch(errors.HandledTunnelingError, err =>
+				this.logger.crit(`Tunneling Error (${err.message})`),
+			)
+			.catch(errors.InvalidHostnameError, () =>
+				cltSocket.end('HTTP/1.0 403 Forbidden\r\n\r\n'),
+			)
+			.catch((err: Error) => {
+				captureException(
+					err,
+					`unknown error establishing tunnel to ${req.url} (${err.message})`,
+					{
+						req,
+					},
+				);
+				cltSocket.end('HTTP/1.0 500 Internal Server Error\r\n\r\n');
+			});
+
+	private forwardRequest = (
+		vpnHost: string,
+		uuid: string,
+		port: number,
+		proxyAuth?: Buffer,
+	): Bluebird<net.Socket> =>
+		new Bluebird((resolve, reject) => {
+			let tunnelProxyResponse = '';
+			const socket: net.Socket = net.connect(3128, vpnHost, () => {
+				socket.write(`CONNECT ${uuid}.balena:${port} HTTP/1.0\r\n`);
+				if (proxyAuth != null) {
+					socket.write(
+						`Proxy-Authorization: Basic ${proxyAuth.toString('base64')}\r\n`,
+					);
+				}
+				socket.write('\r\n\r\n');
+			});
+
+			const earlyEnd = () => {
+				reject(
+					new errors.RemoteTunnellingError(
+						`could not connect to device ${uuid} on port ${port}: tunneling socket closed prematurely.`,
+					),
+				);
+			};
+			const earlyError = (err: Error) => {
+				let errMsg = 'could not connect to vpn tunnel';
+				if (err != null && err.message) {
+					errMsg += `: ${err.message}`;
+				}
+				captureException(err, errMsg);
+				reject(new errors.RemoteTunnellingError(errMsg));
+			};
+			const proxyData = (chunk: Buffer) => {
+				if (chunk != null) {
```

Comment:
You can change this to
```ts
if (chunk == null) {
	return;
}
tunnelProxyResponse += chunk.toString();
```
because you know that the
```ts
if (!_.includes(tunnelProxyResponse, '\r\n\r\n')) {
```
below won't change if there's no chunk.

That said is it even possible for the chunk to be null?

Reply:
> That said is it even possible for the chunk to be null?

Hmm, I think not.